### ISSUE TITLE
SqlServerSecureConnection: Restart Service Timeout

### ISSUE THREAD

**Issue by Teutenberg** [PowerShell/SqlServerDsc#1306]
#### Details of the scenario you tried and the problem that is occurring
Set method fails due to service restart timeout. 120 seconds is the default timeout and cannot be changed. I have checked and the service does successfully restart, so it must be taking longer to restart than the timeout allows. And after the failure the Test method returns InDesiredSate = True, however Configuration Manager doesn't show the certificate as configured.  The Get method shows the correct certificate thumbprint, the registry and wmi show it as correct too.

#### Verbose logs showing the problem
VERBOSE: [SPDEVD01]:
[[SqlServerSecureConnection]DirectResourceAccess] Connected to SQL instance
'localhost\SP2013'.
VERBOSE: [SPDEVD01]:
[[SqlServerSecureConnection]DirectResourceAccess] Getting information about
service 'SQL Server'.
VERBOSE: [SPDEVD01]:
[[SqlServerSecureConnection]DirectResourceAccess] 'SQL Server' service is
restarting.
VERBOSE: [SPDEVD01]:
[[SqlServerSecureConnection]DirectResourceAccess] Starting service 'SQL Server
Agent (SP2013)'. March 15th 2019 09:56:01Info
VERBOSE: [SPDEVD01]:
[[SqlServerSecureConnection]DirectResourceAccess] Waiting for instance
localhost\SP2013 to report status online, with a timeout value of 120 seconds.
VERBOSE: [SPDEVD01]:
[[SqlServerSecureConnection]DirectResourceAccess] Found PowerShell module
SqlServer already imported in the session.
VERBOSE: [SPDEVD01]: LCM:  [ End    Set      ]
[[SqlServerSecureConnection]DirectResourceAccess]  in 27.9840 seconds.
PowerShell DSC resource MSFT_SqlServerSecureConnection  failed to execute
Set-TargetResource functionality with error message:
System.InvalidOperationException: Failed to connect to the instance
localhost\SP2013 within the timeout period of 120 seconds.
VERBOSE: Operation 'Invoke CimMethod' complete.
    + CategoryInfo          : InvalidOperation: (root/Microsoft/...gurationMan
   ager:String) [], CimException
    + FullyQualifiedErrorId : ProviderOperationExecutionFailure
    + PSComputerName        : localhost
VERBOSE: Time taken for configuration job to complete is 28.181 seconds 

#### Suggested solution to the issue
Increase Timeout or pass param SkipWaitForOnline 

#### The DSC configuration that is used to reproduce the issue (as detailed as possible)
  $Params = @{
      InstanceName    = "SP2013"
      Thumbprint      = "‎bf035307f6dd5198832304f0e2f874fc6fda1670"
      ForceEncryption = $false
      Ensure          = 'Present'
      ServiceAccount  = "DOM\SrvSQLsrvSP01dev"
  }

  $Test = Invoke-DscResource -ModuleName SqlServerDsc -Name SqlServerSecureConnection -Property $Params -Method Test

  if (!$Test) {
      Write-Host "Configuring(SqlServerSecureConnection)... $SQLServer"
      Invoke-DscResource -ModuleName SqlServerDsc -Name SqlServerSecureConnection -Property $Params -Method Set -Verbose
  }
  else {
      Write-Host "Skipping(SqlServerSecureConnection)... SQL Instance already configured. $SQLServer"
  }

#### SQL Server edition and version the target node is running
SQL Server 2014 Enterprise

#### SQL Server PowerShell modules present on the target node
Name         Version    Path                                                                                                   
----         -------    ----                                                                                                   
SqlServer    21.1.18068 C:\Program Files\WindowsPowerShell\Modules\SqlServer\21.1.18068\SqlServer.psd1                         
SQLASCMDLETS 1.0        C:\Program Files (x86)\Microsoft SQL Server\120\Tools\PowerShell\Modules\SQLASCMDLETS\SQLASCMDLETS.psd1
SQLPS        1.0        C:\Program Files (x86)\Microsoft SQL Server\120\Tools\PowerShell\Modules\SQLPS\SQLPS.psd1  

#### The operating system the target node is running
OsName               : Microsoft Windows Server 2012 R2 Standard
OsOperatingSystemSKU : StandardServerEdition
OsArchitecture       : 64-bit
WindowsBuildLabEx    : 9600.19304.amd64fre.winblue_ltsb_escrow.190305-1818
OsLanguage           : en-US
OsMuiLanguages       : {en-US}

#### Version and build of PowerShell the target node is running
Name                           Value                                                                                                                                                                                 
----                           -----                                                                                                                                                                                 
PSVersion                      5.1.14409.1018                                                                                                                                                                        
PSEdition                      Desktop                                                                                                                                                                               
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                                                                                                               
BuildVersion                   10.0.14409.1018                                                                                                                                                                       
CLRVersion                     4.0.30319.42000                                                                                                                                                                       
WSManStackVersion              3.0                                                                                                                                                                                   
PSRemotingProtocolVersion      2.3                                                                                                                                                                                   
SerializationVersion           1.1.0.1 

#### Version of the DSC module that was used ('dev' if using current dev branch)
SqlServerDsc 12.1.0.0 C:\Program Files\WindowsPowerShell\Modules\SqlServerDsc\12.1.0.0\SqlServerDsc.psd1

**Comment by johlju** [comment]
I agree the timeout should be able to be controlled. Maybe we could add both options, control the timeout or skip with `SkipWaitForOnline`.

Could you test if it is the timeout? On this line

https://github.com/PowerShell/SqlServerDsc/blob/d36a171abc7b4ab6cba1f49c47742d7a496e9eae/DSCResources/MSFT_SqlServerSecureConnection/MSFT_SqlServerSecureConnection.psm1#L242

Can you change that line to this and run again?

```powershell
Restart-SqlService -SQLServer localhost -SQLInstanceName $InstanceName -Timeout 300
```powershell

Change the value until it works, so we know what a good default value should be for a timeout for this resource.

**Comment by johlju** [comment]
> however Configuration Manager doesn't show the certificate as configured. The Get method shows the correct certificate thumbprint, the registry and wmi show it as correct too.

Does it need a restart to re-read cached values? If this is not the case, then pleaseopen another issue so we can track this problem.

**Comment by Teutenberg** [comment]
Doesn't look like anything changes when upping the Timeout value. Says it failed to connect within the timeout period of 1200 but the job only took 29.679 seconds....


VERBOSE: [SPDEVD01]:
March 18th 2019 15:12:28Info
[[SqlServerSecureConnection]DirectResourceAccess] Waiting for instance  
March 18th 2019 15:12:28Info
localhost\SP2013 to report status online, with a timeout value of 1200 seconds. 
March 18th 2019 15:12:28Info
VERBOSE: [SPDEVD01]:                             
March 18th 2019 15:12:28Info
[[SqlServerSecureConnection]DirectResourceAccess] Found PowerShell module  
March 18th 2019 15:12:28Info
SqlServer already imported in the session. 
March 18th 2019 15:12:38Info
VERBOSE: [SPDEVD01]: LCM:  [ End    Set      ]   
March 18th 2019 15:12:38Info
[[SqlServerSecureConnection]DirectResourceAccess]  in 29.4590 seconds. 
March 18th 2019 15:12:38Error
PowerShell DSC resource MSFT_SqlServerSecureConnection  failed to execute  
March 18th 2019 15:12:38Error
Set-TargetResource functionality with error message:  
March 18th 2019 15:12:38Error
System.InvalidOperationException: Failed to connect to the instance  
March 18th 2019 15:12:38Error
localhost\SP2013 within the timeout period of 1200 seconds.  
March 18th 2019 15:12:38Error
    + CategoryInfo          : InvalidOperation: (root/Microsoft/...gurationMan  
March 18th 2019 15:12:38Error
   ager:String) [], CimException 
March 18th 2019 15:12:38Error
    + FullyQualifiedErrorId : ProviderOperationExecutionFailure 
March 18th 2019 15:12:38Info
VERBOSE: Operation 'Invoke CimMethod' complete. 
March 18th 2019 15:12:38Error
    + PSComputerName        : localhost 
March 18th 2019 15:12:38Error
 
March 18th 2019 15:12:38Info
VERBOSE: Time taken for configuration job to complete is 29.679 seconds

**Comment by Teutenberg** [comment]
Completes successfully when passing parameter SkipWaitForOnline. Must be bug in Restart-SqlService?

Restart-SqlService -SQLServer localhost -SQLInstanceName $InstanceName -SkipWaitForOnline $True

**Comment by Teutenberg** [comment]
Looking at the code, I'm guessing Restart-SqlService Timeout loop is calling Connect-Sql which is throwing the InvalidOperationException. The Variable $testConnectionServerObject is then null, and the TimeOut loop hits the break.

**Comment by johlju** [comment]
@Teutenberg Do you feel we can close this one after we merged PR #1376?

**Comment by Teutenberg** [comment]
Yes this can be closed. I have added a Pester test for Restart-SqlService to check it waits for timeout 10 seconds. PR#1384
